futures-util/src/stream/stream/flatten_unordered.rs

```diff
@@ -0,0 +1,388 @@
+        self.state
+            .fetch_update(Ordering::SeqCst, Ordering::SeqCst, |state| {
+                if state & WAKING_ANYTHING == NONE {
+                    Some(POLLING)
+                } else {
+                    None
+                }
+            })
+            .ok()
+    }
+
+    /// Starts the waking process and performs bitwise or with the given value.
+    fn start_waking(&self, to_poll: u8, waking: u8) -> u8 {
+        self.state.fetch_or(to_poll | waking, Ordering::SeqCst)
+    }
+
+    /// Toggles state to non-waking, allowing to start polling.
+    fn stop_waking(&self, waking: u8) -> u8 {
+        self.state.fetch_and(!waking, Ordering::SeqCst)
+    }
+
+    /// Sets current state to `!POLLING`, allowing to use wakers.
+    fn stop_polling(&self, to_poll: u8) -> u8 {
+        self.state
+            .fetch_update(Ordering::SeqCst, Ordering::SeqCst, |state| {
+                Some((state | to_poll) & !POLLING)
+            })
+            .unwrap()
+    }
+}
+
+/// Waker which will update `poll_state` with `need_to_poll` value on
+/// `wake_by_ref` call and then, if there is a need, call `inner_waker`.
+struct InnerWaker {
+    inner_waker: UnsafeCell<Option<Waker>>,
+    poll_state: SharedPollState,
+    need_to_poll: u8,
+}
+
+unsafe impl Send for InnerWaker {}
+
+unsafe impl Sync for InnerWaker {}
+
+impl InnerWaker {
+    /// Replaces given waker's inner_waker for polling stream/futures which will
+    /// update poll state on `wake_by_ref` call. Use only if you need several
+    /// contexts.
+    ///
+    /// ## Safety
+    ///
+    /// This function will modify waker's `inner_waker` via `UnsafeCell`, so
+    /// it should be used only during `POLLING` phase.
+    unsafe fn replace_waker(self_arc: &mut Arc<Self>, cx: &Context<'_>) -> Waker {
+        *self_arc.inner_waker.get() = cx.waker().clone().into();
+        waker(self_arc.clone())
+    }
+
+    // Flags state that walking is started for the walker with the given value.
+    fn start_waking(&self) -> u8 {
+        self.poll_state.start_waking(self.need_to_poll, self.need_to_poll << 3)
+    }
+
+    // Flags state that walking is finished for the walker with the given value.
+    fn stop_waking(&self) -> u8 {
+        self.poll_state.stop_waking(self.need_to_poll << 3)
+    }
+}
+
+impl ArcWake for InnerWaker {
+    fn wake_by_ref(self_arc: &Arc<Self>) {
+        let poll_state_value = self_arc.start_waking();
+
+        // Only call waker if stream isn't being polled because of safety reasons.
+        // Waker will be called at the end of polling if state was changed.
+        if poll_state_value & POLLING == NONE {
+            if let Some(inner_waker) =
+                unsafe { self_arc.inner_waker.get().as_ref().cloned().flatten() }
+            {
+                // First, stop waking to allow polling stream
+                self_arc.stop_waking();
+                // Wake inner waker
+                inner_waker.wake();
```

Comment:
If waker panics, the state will be fixed to POLLING and it makes FlattenUnordered unusable, right?

Reply:
If waker panics, everything will go well as `self_arc.stop_waking()` was called prior to calling a waker, see https://github.com/rust-lang/futures-rs/pull/2083/commits/9bc4e4eedcc6e98218346d688cac2d2a3a4962b7

However, after any `poll_next`'s panic state was corrupted, so I added some state bombs to prevent this behavior.